gquant/dataframe_flow/_node_flow.py

```diff
@@ -138,61 +153,155 @@ def get_pout(df_out):
+                    ioport=iport,
+                    ioport_types=iport_types)
+
+                msgo = msgfmt.format(
+                    task=from_node.uid,
+                    nodetype=_get_nodetype(from_node),
+                    inout='output',
+                    ioport=oport,
+                    ioport_types=oport_types)
+
+                errmsg = 'Port Types Validation\n{}\n{}\n'\
+                    'Connected nodes do not have matching port types. Fix '\
+                    'port types.'.format(msgi, msgo)
+
+                raise TypeError(errmsg)
+
     def validate_required_columns(self):
 
-        def validate_required(icols, kcol, kval, in_taskid=None, iport=None):
+        def validate_required(iport, kcol, kval, ientnode, icols):
+            node = ientnode['from_node']
+            oport = ientnode['from_port']
+            src_task = '{}.{}'.format(node.uid, oport)
+            src_type = _get_nodetype(node)
+            # incoming "task.port":"Node-type":{{column:column-type}}
+            msgi = \
+                '"{task}":"{nodetype}" produces columns {colinfo}'.format(
+                    task=src_task,
+                    nodetype=src_type,
+                    colinfo=icols)
+
+            dst_task = '{}.{}'.format(self.uid, iport)
+            dst_type = _get_nodetype(self)
+            # expecting "task.port":"Node-type":{{column:column-type}}
+            msge = \
+                '"{task}":"{nodetype}" requires columns {colinfo}'.format(
+                    task=dst_task,
+                    nodetype=dst_type,
+                    colinfo={kcol: kval})
+
+            header = \
+                'Columns Validation\n'\
+                'Format "task.port":"Node-type":{{column:column-type}}'
+            info_msg = '{}\n{}\n{}'.format(header, msgi, msge)
+
             if kcol not in icols:
-                err_msg = 'Incoming columns not valid: error for node "%s", '\
-                    'missing required column "%s".' % (self.uid, kcol)
-                if in_taskid:
-                    dst_uid = self.uid if iport is None else \
-                        '{}.{}'.format(self.uid, iport)
-                    err_msg = '{}\nIncoming columns from "{}" do not match '\
-                        'columns_setup for "{}".'.format(
-                            err_msg, in_taskid, dst_uid)
-                # raise Exception(err_msg)
-                print(err_msg)
-            if (kcol in icols) and kval != icols[kcol]:
+                err_msg = \
+                    'Task "{}" missing required column "{}" '\
+                    'from "{}".'.format(self.uid, kcol, src_task)
+                out_err = '{}\n{}'.format(info_msg, err_msg)
+                raise LookupError(out_err)
+
+            ival = icols[kcol]
+            if kval != ival:
                 # special case for 'date'
-                if (kval == 'date' and icols[kcol]
+                if (kval == 'date' and ival
```

Comment:
Maybe we should extract this out so we people can customize the equavalent type strings like the 'date' one here.

Reply:
Yes, that would be a good idea. I don't want to that in this PR. I'm probably already doing too much.

This code was because of the `cudf.read_csv` functionality. One would specify `date` as column dtype, but the resulting cudf had one of `datetime` types. This logic here was added to deal with this inconsistency of getting a different datatype out than what was specified.

Reply:
Yes. Let's do it for the future